Add live kernel version checks and package status

r0x0d opened the pull request:
This PR aims to add a couple of checks before the conversion runs. The checks will prevent the conversion from falling before the process is finished. 

The checks are the following: 

* Check if all the system packages are updated (Through yum/dnf API calls) in enabled repos
* Check if the booted kernel is at the latest version (Through a comparison between `rpm` and `uname`)

If both of the conditions are met, then the conversion will proceed normally in the workflow.

Ticket reference: https://issues.redhat.com/browse/OAMG-4659
Linked ticket: https://issues.redhat.com/browse/OAMG-4709

abadger commented on convert2rhel/checks.py:
```diff
@@ -449,3 +453,47 @@ def _bad_kernel_substring(kernel_release):
+    # The installed kernel version
+    uname_output, _ = run_subprocess("uname -r")
+
+    match = LooseVersion(repoquery_output) > LooseVersion(uname_output)
```
I see that we `import rpm` in the pkghandler file so we already have a dependency on that.  rpm has a function to compare NEVR (Name epoch version release). `rpm.labelCompare((epoch1, version1, release1), (epoch2, version2, release2))`.  

convert2rhel is GPLv3 or later and rpmdevtools is GPLv2 or later (correct me if I'm wrong on either of those points :-) so you should be able to look at the source of rpmdev-rpmvercmp for some additional examples: https://pagure.io/rpmdevtools/blob/main/f/rpmdev-vercmp#_108

abadger replied:
Note that arch probably doesn't make sense as part of greater than and less than comparisons.  But it would make sense for an equality comparison.  Perhaps you want to do something to make sure that you have the right arch prior to doing your version comparison? (Or you might want to assume correct arch due to having setup the repositories earlier?)

r0x0d replied:
Do you guys think that comparing versions is a really important step on this? I mean, `repoquery` should output the latest version available at the current repositories, so if the `uname -r` doesn't match the output of `repoquery` it would mean that it's not updated, right? 

Don't know if I'm really on the correct train of thought here. If not, I can change to do version comparison with `rpm.labelCompare()`

abadger replied:
I'm not familiar enough to give a definitive answer but here are some ideas of potential problems:

* What do we want to do if the repository is out-of-sync and reports an older version?
* What do we want to do in this function if the user has a non-standard but newer kernel installed?

r0x0d replied:
In a recent call with @bocekm we were able to get this right. I've added the code for properly handling the `repoquery` and `uname` output, and to add the `rpm.labelCompare` function too.

abadger commented on convert2rhel/checks.py:
```diff
@@ -449,3 +453,47 @@ def _bad_kernel_substring(kernel_release):
+    # There is a whole blank line before outputting the packages to update
+    # and to output the real counting, we just -1 the counting.
+    packages_to_update_count = yum_stdout.count("\n") - 1
+
```
It would be slower but more robust to split the output and filter out the lines we don't want.  I also noticed that when I ran yum check-update -q locally, I got other, non package lines:
> vim-minimal.x86_64                             2:8.2.3568-1.fc33                        updates-testing
> Security: kernel-core-5.14.15-100.fc33.x86_64 is an installed security update
> Security: kernel-core-5.14.13-100.fc33.x86_64 is the currently running version

What do you think about:

``` python
lines = yum_stdout.split("\n")

# Filter lines that only have whitespace or are blank
lines = (line for line in yum_stdout.split("\n") if line.strip())
# Filter informational lines that don't have to do with available packages
lines = (line for line in lines if not line.endswith("is an installed security update"))
packages = [line for line in lines if not line.endswith("is the currently running version")]

packages_to_update_count = len(packages)
```

r0x0d replied:
I agree with @SpyTec, the code looks way cleaner and easy to understand with the regex. I tested and got all the packages that needed to update in one go, and if we have to output to the user in the future about the packages, we can just iterate over the list and show the names to them.

abadger replied:
<nod> Regexes can simplify surrounding code.  However, they mean you have a regex to maintain and understand. So do be mindful of that.

https://www.explainxkcd.com/wiki/index.php/1171:_Perl_Problems

bocekm replied:
The regex would not handle well the situation with the security update @abadger mentioned - it would give us two packages to update even though there's none. So, I'm for the @abadger's approach.

bocekm replied:
@j-mracek, in the output of `yum check-update` shared by @abadger in https://github.com/oamg/convert2rhel/pull/356/files#r741452001 (`is an installed security update`/`is the currently running version`) we can see the same NEVRA at both lines. I suppose the NEVRAs can differ, right? Would in such a case such NEVRA appear with the "Security:" prefix or would it appear on a separate line without the prefix as well?

Also, is there any yum/dnf python API function that would give us the information provided by the `check-update` CLI command?

j-mracek replied:
I am little bit lost here, but I will try my best. There are multiple possibilities how to receive information provided by check-update command. Try use Query class with filters updates, and security. Please can you describe exactly what set of packages are you looking for? Then I can look in implementation.

r0x0d replied:
@j-mracek We are looking for general packages updates. We are not aiming at a specific set of packages, but rather, if there is any package to be updated in the system. Since @abadger provided us with a different output from the ones I've got in my testing, we were wondering if `yum check-update` could give us any different messages that we are not aware of.

j-mracek replied:
What about:
In case that you can directly play with goal in base
base.upgrade_all()
base.resolve()
Then iterate over base.transaction
for x in base.transaction:
    print(x, x.action_name)

r0x0d replied:
@j-mracek For some reason, I was not able to manage to get your code snippet working. It doesn't output anything for me when I run it, maybe I'm doing something wrong? 

```python
import dnf

db = dnf.Base()

db.fill_sack()

db.upgrade_all()
db.resolve()

[print(package) for package in db.transaction]
```

I also tried like this: 

```python
import dnf
db = dnf.Base()

db.fill_sack()

query = db.sack.query()
installed = query.upgrades()

print("Installed dnf package:")
for pkg in list(installed):
    print(pkg, pkg.reponame)
```

But none of the above worked... Already checked the output of `dnf check-update` command and it reported I have 50 packages to update.

Also, @j-mracek, I've checked the code for dnf/yum regarding the command `check-update` and both seem to use a different method. Do you think we could also try to do to the same as the tool? Or is the solution you shared with us easier to implement?

Thanks in advance for your help!

j-mracek replied:
Do you still need any assistance?

r0x0d replied:
@j-mracek, for now, I think I'm all good! Thanks for your helping and patience!

SpyTec commented on convert2rhel/unit_tests/checks_test.py:
```diff
@@ -813,3 +825,86 @@ def test_custom_repos_are_invalid(self):
+    )
+
+    if raise_system_exit:
+        with pytest.raises(SystemExit):
+            assert check_package_updates()
```
After the raises, it would be good to check if the amount of packages is accurate. `The system has %s packages to be updated`

r0x0d replied:
Updated that assert as well.

SpyTec commented on convert2rhel/unit_tests/checks_test.py:
```diff
@@ -813,3 +825,86 @@ def test_custom_repos_are_invalid(self):
+    ("package", "return_code", "raise_system_exit"),
+    (
+        (
+            "convert2rhel.noarch   0.24-1.20211019110627581338.pr343.11.g4ff8753.el8   copr:copr.fedorainfracloud.org:group_oamg:convert2rhel\n"
+            "convert2rhel.src      0.24-1.20211019110627581338.pr343.11.g4ff8753.el8   copr:copr.fedorainfracloud.org:group_oamg:convert2rhel\n"
+            "tzdata.noarch         2021c-1.el8                                         baseos\n",
```
Add this as well since Toshio mentioned it can appear
```
Security: kernel-core-5.14.15-100.fc33.x86_64 is an installed security update
Security: kernel-core-5.14.13-100.fc33.x86_64 is the currently running version
```

r0x0d replied:
@SpyTec, Those two lines are just informational and not telling the end-user that `kernel-core` needs to be updated? So our solution should gracefully ignore that, right?

SpyTec replied:
Yeah exactly. We should still have it in the test to make sure we don't pick up on it though in the function

r0x0d replied:
Added a few more cases to the test as well to the addition of the two lines of the `Security:`log

SpyTec commented on convert2rhel/unit_tests/checks_test.py:
```diff
@@ -813,3 +825,86 @@ def test_custom_repos_are_invalid(self):
+    )
+
+    if raise_system_exit:
+        with pytest.raises(SystemExit):
+            check_installed_kernel_version()
```
Check logs here too if possible to see that the current and latest kernel is correct as well

r0x0d replied:
Added the assserts corresponding to this one.

SpyTec commented on convert2rhel/checks.py:
```diff
@@ -449,3 +453,47 @@ def _bad_kernel_substring(kernel_release):
+
+def check_package_updates():
+    """Ensures that the system packages installed are up-to-date."""
+    logger.task("Prepare: Checking if the packages are up-to-date")
```
```suggestion
    logger.task("Prepare: Checking if the system packages are up-to-date")
```

SpyTec commented on convert2rhel/checks.py:
```diff
@@ -449,3 +453,47 @@ def _bad_kernel_substring(kernel_release):
+    logger.task("Prepare: Checking if the installed kernel version is the most recent.")
+
+    # The latest kernel version on repo
+    repoquery_output, _ = run_subprocess('repoquery --qf "%{version}-%{release}.%{arch}" kernel')
```
This could show multiple kernel versions, at least does for me on Fedora. Will this happen on CentOS or Oracle Linux?

r0x0d replied:
@SpyTec, Could you provide the output for me? I think it may be getting kernels from other repositories... If so, this is something I haven't tested. Trying to do some more examples to catch those kernels. 

I'm thinking that we can use the same approach as here: https://github.com/oamg/convert2rhel/pull/353/files#diff-cf19bc90f65d9e26e1b6fa7dcd84a18dfba097f65706e211a4f4ac71b1b467f6R389

We could include the BUILDTIME variable and sort it out for the newest one, what do you think?

abadger replied:
Here's output on my Fedora 33 system:

> [pts/15@camelback /srv/convert2rhel/c2r-vagrant]$ repoquery --qf "%{version}-%{release}.%{arch}" kernel 
> Last metadata expiration check: 3:01:15 ago on Thu 04 Nov 2021 08:32:35 AM PDT.
> 5.14.15-100.fc33.x86_64
> 5.8.15-301.fc33.x86_64


I think it might retrieve the latest package version for each repository?  In Fedora there's a repository for the release and separate repositories for updates so the kernel package could appear in all three of those.

I'm not sure if it's the same with rhel or if they simply have a monolithic repo.

r0x0d replied:
I see @abadger, gonna research a bit more on the `repoquery` output. I think it's querying on all repos the user may have enabled in his system, so it indeed could be a problem. The easiest way I can think of fixing this is sorting out the latest `kernel` as I did previously in another PR. 

Thanks for your output!

SpyTec reviewed (dismissed):
Overall code looks good!

abadger commented on convert2rhel/checks.py:
```diff
@@ -30,6 +32,7 @@
 
 logger = logging.getLogger(__name__)
 
+PACKAGE_NAMES_RE = r"^[0-9]*:?([a-z][a-z0-9-_]*?)(-[0-9])*\."
```
One problem with regexes is that they are generally slower than other ways of checking.  Assigning regexes outside of any potential code loops like you're doing here is one way to mitigate that.  However, you need to not just assign the string here but also compile the regex (the speed savings is because you only have to compile the regex once, when the module is first imported).   So do something like this:
```suggestion
PACKAGE_NAMES_RE = re.compile(r"^[0-9]*:?([a-z][a-z0-9-_]*?)(-[0-9])*\.")
```
```

abadger replied:
Hmm.. also... does this regex work for the data we're being given?  I was going to make a suggestion about the other problem with regexes being that they're hard to debug so we should comment new regexes heavily.  But when I tried running some yum check-output data through it so I could understand what this regex is doing, I came out with zero matches instead of the one match I was expecting.

``` python 
>>> import re
>>> import sh  # amoffat's sh library
>>> a = re.compile(r"^[0-9]*:?([a-z][a-z0-9-_]*?)(-[0-9])*\.")
>>> yum_stdout = sh.yum(['check-update', '-q'], _ok_code=[0, 100])
>>> yum_stdout

python3-pyOpenSSL.noarch              21.0.0-1.fc33              updates-testing
Security: kernel-core-5.14.15-100.fc33.x86_64 is an installed security update
Security: kernel-core-5.14.13-100.fc33.x86_64 is the currently running version

>>> re.findall(a, str(yum_stdout))
[]
```

r0x0d replied:
@abadger I've tested with the 2 ways, and now (yesterday I got the same output as you, an empty []) they are working as expected. I've changed to your suggestions (compiling the regex first) 

```python
('packages', [(u'binutils', u''), (u'convert2rhel', u''), (u'glibc', u''), (u'glibc-common', u''), (u'grub2', u''), (u'grub2-common', u''), (u'grub2-pc', u''), (u'grub2-pc-modules', u''), (u'grub2-tools', u''), (u'grub2-tools-extra', u''), (u'grub2-tools-minimal', u''), (u'kernel', u''), (u'kernel-tools', u''), (u'kernel-tools-libs', u''), (u'nfs-utils', u''), (u'nspr', u''), (u'nss', u''), (u'nss-softokn', u''), (u'nss-softokn-freebl', u''), (u'nss-sysinit', u''), (u'nss-tools', u''), (u'nss-util', u''), (u'openssl', u''), (u'openssl-libs', u''), (u'python-perf', u''), (u'sudo', u''), (u'tzdata', u'')])
```

And the log message

```bash
CRITICAL - The system has 27 packages to be updated. To proceed with the conversion, update the packages on your system by executing the following step:
```

abadger replied:
Interesting.  Can you post your output from `yum check-update` or explain how the regex works?  I'm seeing that the unittests also appear to be failing in the same way as my local attempts are... zero packages are found in the output.

r0x0d replied:
As @bocekm stated in another thread here, we decided to go with the manual approach for that validation as it is handling better in some other cases for `yum check-update -q` command. 

And as far as the regex, I don't really know why the outputs were different. I'm not an expert on regex, but I think it was something along with the regex compilation method. Don't really know what caused the empty `[]` on the output.

abadger commented on convert2rhel/checks.py:
```diff
@@ -447,3 +452,57 @@ def _bad_kernel_substring(kernel_release):
+    lines = (line for line in yum_stdout.split("\n") if line.strip())
+    # Filter informational lines that don't have to do with available packages
+    lines = (line for line in lines if not line.endswith("is an installed security update"))
+    packages = [line for line ingit lines if not line.endswith("is the currently running version")]
```
```suggestion
    packages = [line for line in lines if not line.endswith("is the currently running version")]
```

r0x0d replied:
Minor typo, fixed!

abadger commented on convert2rhel/unit_tests/checks_test.py:
```diff
@@ -813,3 +825,89 @@ def test_custom_repos_are_invalid(self):
+    ("package", "return_code", "raise_system_exit"),
+    (
+        (
+            "convert2rhel.noarch   0.24-1.20211019110627581338.pr343.11.g4ff8753.el8   copr:copr.fedorainfracloud.org:group_oamg:convert2rhel\n"
```
```suggestion
            "\n",
            "convert2rhel.noarch   0.24-1.20211019110627581338.pr343.11.g4ff8753.el8   copr:copr.fedorainfracloud.org:group_oamg:convert2rhel\n"
```

Also have an initial blank line since that's also a feature of the yum check-updates output

r0x0d replied:
Got it. Added to the tests case.

r0x0d commented on convert2rhel/checks.py:
```diff
@@ -447,3 +452,57 @@ def _bad_kernel_substring(kernel_release):
+    total_packages = len(packages)
+
+    if total_packages > 0:
+        logger.critical(
```
We could change this to a `logger.warning()` instead of the critical. I think the `warning` will not cause an `SystemExit`, right?

r0x0d commented on convert2rhel/checks.py:
```diff
@@ -447,3 +452,57 @@ def _bad_kernel_substring(kernel_release):
+
+def check_installed_kernel_version():
+    """Check if the installed kernel is behind or at the same version in yum repos"""
+    logger.task("Prepare: Checking if the installed kernel version is the most recent.")
```
Changed to the more appropriate function name

kokesak commented:
After the first discussion with @r0x0d  we created a test case table with scenarios on how to test this PR.
https://docs.google.com/spreadsheets/d/17fw2qFEUI9g-M5_sf3uFnwIVhn7zswcj0eETvEOKSD0/edit#gid=0

This table will be updated as the development on this task progress.

thrix commented:
/packit test

r0x0d commented on convert2rhel/pkghandler.py:
```diff
@@ -895,3 +894,51 @@ def compare_package_versions(version1, version2):
+
+    packages = base.returnPkgLists(["updates"], None)
+
+    print("PACKAGES: ", packages)
```
It's not done yet. The print is just for debugging purposes, will remove it when it's done

pirat89 commented on convert2rhel/checks.py:
```diff
@@ -447,3 +456,57 @@ def _bad_kernel_substring(kernel_release):
+    logger.task("Prepare: Checking if the loaded kernel version is the most recent.")
+
+    # The latest kernel version on repo
+    repoquery_output, _ = run_subprocess('repoquery -q --qf "%{BUILDTIME}\\t%{VERSION}-%{RELEASE}" kernel')
```
@r0x0d @oamg/convert2rhel-developers   what in case the repository is not reachable? e.g.
* detached systems
* disabled repositories

in such a case you will not obtain any information to be able to compare the versions. (we already hit such a systems multiple times). unless you have written in requirements to have enabled base repositories for the particular system (for OL, CentOS, ..). If not, I suggest to have a fallback to check the latest installed kernel as well.

Additionally, for the installed packages, for EL 8+ you need to care about `kernel-core` rpms instead of `kernel`, as the `kernel` rpm is just a metapackage since EL 8 and it doesn't have to be installed at all (this is not problem when you check what's newest pkg in repository - if repo is enabled and accessible)

pirat89 replied:
to simulate such a situation, here is my output from centos 7 container
```
[root@1c732f612637 /]# repoquery -q --qf "%{BUILDTIME}\\t%{VERSION}-%{RELEASE}" --disablerepo="*" kernel
[root@1c732f612637 /]# echo $?
0
```

r0x0d replied:
Hey, @pirat89! Thanks for your insights on this. Just out of curiosity, how did you guys handle those cases you mentioned? I don't recall having a programmatic set of enabling repositories in the tool, as far as I know, having at least the base repositories enabled is a user requirement (But I may be wrong on that). 

@SpyTec do you know any cases where the base repos for a system are disabled for our users?

SpyTec replied:
@pirat89 From our stand-up and GChat, we require repos to be enabled as mentioned in the docs. So not an issue we need to worry about for this PR

> Access to repositories providing packages from the original distribution is required, especially to the base repository, to perform a rollback. 

https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html-single/converting_from_an_rpm-based_linux_distribution_to_rhel/index#con_rollback_converting-from-a-linux-distribution-to-rhel

pirat89 replied:
Hi guys, sorry for late response. I am not looking here so much during my holiday. I've read the discussion and it seems you should not have any problems here regarding the hard requirement to have enabled the base repository on the system. The unsupported envar seems as a good decision to bypass that requirement, keeping the possibility to handle problems by customers on their own.

@r0x0d  I spoke about cases we discovered in case of in-place upgrades between major versions of RHEL systems - probably I should mention that, sorry for the confusion. I am trying to inform you about problems we already had to deal with when I see you are dealing with similar problems :)

bocekm commented on convert2rhel/checks.py:
```diff
@@ -447,3 +456,59 @@ def _bad_kernel_substring(kernel_release):
+    package_to_check = "kernel-core" if system_info.version.major >= 8 else "kernel"
+
+    # The latest kernel version on repo
+    repoquery_output, _ = run_subprocess('repoquery -q --qf "%{BUILDTIME}\\t%{VERSION}-%{RELEASE}" ' + package_to_check)
```
```suggestion
    repoquery_output, _ = run_subprocess('repoquery --quiet --qf "%{BUILDTIME}\\t%{VERSION}-%{RELEASE}" ' + package_to_check)
```
The `-q` means `no-op for rpmquery compatibility` on RHEL 6 - i.e. it does nothing.

r0x0d commented on requirements/centos8.requirements.txt:
```diff
@@ -1,5 +1,5 @@
 pylint==2.6.0
 astroid==2.4.2
-pytest==6.2.1
+pytest==6.2.5
```
pytest was throwing an error when running locally with python `3.10.0`, the fix was upgrading the version to `6.2.5`

r0x0d commented on convert2rhel/pkgmanager.py:
```diff
@@ -14,10 +14,16 @@
 try:
     from yum import *
 
+    # Yum's installation path for the CLI code
+    # https://github.com/rpm-software-management/yum/blob/4ed25525ee4781907bd204018c27f44948ed83fe/bin/yum#L26
+    sys.path.insert(0, "/usr/share/yum-cli")
```
This will be imported every time `yum` is detected, should I move it to here? https://github.com/oamg/convert2rhel/blob/11bab92119510bb40eb7e9c9bc9d0010f7122ff7/convert2rhel/pkghandler.py#L916 For now, this line is only used by this function.

bocekm replied:
As it is now the `sys.path.insert(0, "/usr/share/yum-cli")` will be executed just once because the pkgmanager.py is loaded just once per convert2rhel execution. But moving it to the pkghandler.py makes sense from the point of view that it logically relates just to the one `_get_packages_to_update_yum()` function.

r0x0d commented on convert2rhel/unit_tests/pkghandler_test.py:
```diff
@@ -1387,6 +1398,79 @@ def test_fix_default_kernel_with_no_incorrect_kernel(self):
+    assert get_total_packages_to_update() == len(packages)
+
+
+@pytest.mark.skipif(pkgmanager.TYPE != "yum", reason="No yum module detected on the system, skipping it.")
+@pytest.mark.parametrize(("packages"), ((["package-1", "package-2", "package-3"],)))
+def test_get_packages_to_update_yum(packages, monkeypatch):
+    from cli import YumBaseCli  # pylint: disable=E0401
+
+    class PkgUpdates:
+        def __init__(self, packages):
+            self.updates = packages
+
+    pkg_lists_mock = mock.Mock(return_value=PkgUpdates(packages))
+
+    monkeypatch.setattr(YumBaseCli, "returnPkgLists", value=pkg_lists_mock)
+
+    assert _get_packages_to_update_yum() == packages
+
+
+@pytest.mark.skipif(pkgmanager.TYPE != "dnf", reason="No dnf module detected on the system, skipping it.")
+@pytest.mark.parametrize(("packages"), ((["package-1", "package-2", "package-3"],)))
+def test_get_packages_to_update_yum(packages, monkeypatch):
+    from dnf import Base  # pylint: disable=E0401
+
+    dummy_mock = mock.Mock()
+
+    monkeypatch.setattr(Base, "read_all_repos", value=dummy_mock)
+    monkeypatch.setattr(Base, "fill_sack", value=dummy_mock)
+    monkeypatch.setattr(Base, "upgrade_all", value=dummy_mock)
+    monkeypatch.setattr(Base, "resolve", value=dummy_mock)
+    monkeypatch.setattr(Base, "transaction", value=packages)
+
+    assert _get_packages_to_update_dnf() == packages
```
With this `pytest.mark.skipif` our coverage decreases a little because there will not be both of the package managers at the same time on a system. 

I know there is a way to combine multiple coverage runs to make it better, but don't know how much effort will be to adapt our CI to do that.

r0x0d replied:
One thing that could be possible is splitting the `test_get_total_packages_to_update` test and reapplying these specific monkeypatchs to be compatible with either dnf or yum when running, so it could be something like this: 

* test_get_total_packages_to_update_with_yum
* test_get_total_packages_to_update_with_dnf

SpyTec replied:
Decrease in coverage is fine, code coverage is only a metric - not a goal

You should rename the method though, there are two tests methods with the same name 

```suggestion
@pytest.mark.skipif(pkgmanager.TYPE != "yum", reason="No yum module detected on the system, skipping it.")
@pytest.mark.parametrize(("packages"), ((["package-1", "package-2", "package-3"],)))
def test_get_packages_to_update_yum(packages, monkeypatch):
    from cli import YumBaseCli  # pylint: disable=E0401

    class PkgUpdates:
        def __init__(self, packages):
            self.updates = packages

    pkg_lists_mock = mock.Mock(return_value=PkgUpdates(packages))

    monkeypatch.setattr(YumBaseCli, "returnPkgLists", value=pkg_lists_mock)

    assert _get_packages_to_update_yum() == packages


@pytest.mark.skipif(pkgmanager.TYPE != "dnf", reason="No dnf module detected on the system, skipping it.")
@pytest.mark.parametrize(("packages"), ((["package-1", "package-2", "package-3"],)))
def test_get_packages_to_update_dnf(packages, monkeypatch):
    from dnf import Base  # pylint: disable=E0401

    dummy_mock = mock.Mock()

    monkeypatch.setattr(Base, "read_all_repos", value=dummy_mock)
    monkeypatch.setattr(Base, "fill_sack", value=dummy_mock)
    monkeypatch.setattr(Base, "upgrade_all", value=dummy_mock)
    monkeypatch.setattr(Base, "resolve", value=dummy_mock)
    monkeypatch.setattr(Base, "transaction", value=packages)

    assert _get_packages_to_update_dnf() == packages
```

SpyTec replied:
Decrease in coverage is fine, code coverage is only a metric - not a goal

You should rename the method though, there are two tests methods with the same name 

```suggestion
@pytest.mark.skipif(pkgmanager.TYPE != "yum", reason="No yum module detected on the system, skipping it.")
@pytest.mark.parametrize(("packages"), ((["package-1", "package-2", "package-3"],)))
def test_get_packages_to_update_yum(packages, monkeypatch):
    from cli import YumBaseCli  # pylint: disable=E0401

    class PkgUpdates:
        def __init__(self, packages):
            self.updates = packages

    pkg_lists_mock = mock.Mock(return_value=PkgUpdates(packages))

    monkeypatch.setattr(YumBaseCli, "returnPkgLists", value=pkg_lists_mock)

    assert _get_packages_to_update_yum() == packages


@pytest.mark.skipif(pkgmanager.TYPE != "dnf", reason="No dnf module detected on the system, skipping it.")
@pytest.mark.parametrize(("packages"), ((["package-1", "package-2", "package-3"],)))
def test_get_packages_to_update_dnf(packages, monkeypatch):
    from dnf import Base  # pylint: disable=E0401

    dummy_mock = mock.Mock()

    monkeypatch.setattr(Base, "read_all_repos", value=dummy_mock)
    monkeypatch.setattr(Base, "fill_sack", value=dummy_mock)
    monkeypatch.setattr(Base, "upgrade_all", value=dummy_mock)
    monkeypatch.setattr(Base, "resolve", value=dummy_mock)
    monkeypatch.setattr(Base, "transaction", value=packages)

    assert _get_packages_to_update_dnf() == packages
```

r0x0d commented on convert2rhel/checks.py:
```diff
@@ -45,11 +51,14 @@
     check_readonly_mounts()
     check_rhel_compatible_kernel_is_used()
     check_custom_repos_are_valid()
+    check_package_updates()
+    is_loaded_kernel_latest()
```
This would be the correct order for these checks?

bocekm replied:
Yes. But please move the following from main.py to run before the `perform_pre_checks()` call:
```
loggerinst.task("Prepare: Clear YUM/DNF version locks")
pkghandler.clear_versionlock()
```
The reason is that when [yum versionlock](https://access.redhat.com/solutions/98873) is in place for kernel, `yum update` will not report the kernel to be updated even when a newer version is available in repos.

r0x0d reviewed (commented):
Reviewed.

r0x0d reviewed (commented):
Review

r0x0d reviewed (commented):
Reviewed and fixed a few diffs.

r0x0d commented:
Just a reminder for me (and the reviewers too).

If this PR get merged before the #389, then, @abadger and @Andrew-ang9 you guys have to change the run_subprocess() calls I've added (Or I could push a commit on that branch too)

Otherwise, if this gets merged after the #389, then I have to rebase and modify the run_subprocess() call.

r0x0d commented:
/packit test

j-mracek commented on convert2rhel/checks.py:
```diff
@@ -448,3 +457,81 @@ def _bad_kernel_substring(kernel_release):
+
+    # The latest kernel version on repo
+    repoquery_output, _ = run_subprocess(
+        ["repoquery", "--quiet", "--qf", '"%{BUILDTIME}\\t%{VERSION}-%{RELEASE}"', package_to_check]
```
Why do you use CLI instead of API. What is the targeted systems and in which versions? (RHEL8, 7, Fedora ...)

j-mracek replied:
Parsing outputs from DNF is sensitive to be contaminated by plugins.

j-mracek replied:
_get_packages_to_update_dnf() uses DNF API - therefore repoquery can be replaced by  dnf query-

```
query = base.sack.query().filterm(name=package_to_check)
# to detect running kernel
running_kernel = base.sack.get_running_kernel()

# Then you can try to compare version of packages by direct comparison of  package object. Be aware that running kernel can be not the latest one installed

```

j-mracek replied:
I think that using dnf API will allow to delete a lot of your code handling parsing output and comparison of versions.

r0x0d replied:
IIRC, by the time I developed this code I was not aware if `yum` had an similar API to interact with repoquery as dnf has, that's why I was using the `repoquery` for that stuff, as it was available in every distro we handle (RHEL8 and RHEL7).

I'm currently looking over the `yum` repository again to see if I can find any APIs that could be used here to have the same effect.

j-mracek commented on convert2rhel/checks.py:
```diff
@@ -448,3 +457,81 @@ def _bad_kernel_substring(kernel_release):
+
+    # For OL/CentOS 8 the `kernel` is just a meta package, instead, we check for `kernel-core`.
+    # But for 6 and 7 releases, the correct way to check is using `kernel`.
+    package_to_check = "kernel-core" if system_info.version.major >= 8 else "kernel"
```
It means that the tool will be working in standard RHEL environment with default provided kernels.

kokesak commented:
Hi @r0x0d, I was just briefly looking at the test runs. There are many fails as of now, I suppose we will need to update tests so they correspond to this change.

I saw this error as one of the most one:
http://pastebin.test.redhat.com/1035993
```
CRITICAL - Could not find any kernel from repositories to compare against the loaded kernel.
```
Do we need to have some special repository on the machine or what this check exactly does?

SpyTec commented on convert2rhel/checks.py:
```diff
@@ -448,3 +457,79 @@ def _bad_kernel_substring(kernel_release):
+    """Check if the loaded kernel is behind or of the same version as in yum repos."""
+    logger.task("Prepare: Checking if the loaded kernel version is the most recent.")
+
+    # For OL/CentOS 8 the `kernel` is just a meta package, instead, we check for `kernel-core`.
+    # But for 6 and 7 releases, the correct way to check is using `kernel`.
```
```suggestion
    # For Oracle/CentOS Linux 8 the `kernel` is just a meta package, instead, we check for `kernel-core`.
    # But for 6 and 7 releases, the correct way to check is using `kernel`.
```

SpyTec commented on convert2rhel/pkghandler.py:
```diff
@@ -933,3 +934,94 @@ def has_duplicate_repos_across_disablerepo_enablerepo_options():
+    """Return the total number of packages to update in the system
+
+    It uses both yum/dnf depending on weather they are installed on the system,
+    if it's OL/CentOS 7 or 6, it will use `yum`, otherwise it should use `dnf`.
```
```suggestion
    if it's Oracle/CentOS Linux 7 or 6, it will use `yum`, otherwise it should use `dnf`.
```

SpyTec reviewed (commented):
Looked over again and code looks good

kokesak commented:
@r0x0d I found out the main reason it's failing. There is an issue when the system has two repo files that contains the same information. Eg. run `cp /etc/yum.repos.d/any-repo.repo /etc/yum.repos.d/any-repo-duplicate.repo`.
Here is the failing output from c2r:
https://privatebin.corp.redhat.com/?a3bbb2919317ace9#7EeFqdmJSAyYtYKPCcB26T7t5D7Ts3tGg2GcikXToMfC

Also I have a question regarding this message:
```
WARNING - The system has 1 packages not updated. If you wish to update then before continuing with the conversion, execute the following step:
            out: 
            out: Run: yum update -y
```
Sometimes we get this message, Is there any way that c2r could report the list of the non-updated packages?

r0x0d commented:
> @r0x0d I found out the main reason it's failing. There is an issue when the system has two repo files that contains the same information. Eg. run `cp /etc/yum.repos.d/any-repo.repo /etc/yum.repos.d/any-repo-duplicate.repo` then run the conversion. Here is the failing output from c2r: https://privatebin.corp.redhat.com/?a3bbb2919317ace9#7EeFqdmJSAyYtYKPCcB26T7t5D7Ts3tGg2GcikXToMfC
> 
> Also I have a question regarding this message:
> 
> ```
> WARNING - The system has 1 packages not updated. If you wish to update then before continuing with the conversion, execute the following step:
>             out: 
>             out: Run: yum update -y
> ```
> 
> Sometimes we get this message, Is there any way that c2r could report the list of the non-updated packages?

I will take a look on this error message, thanks for the debugging. 

And regarding showing the list of packages that needs an update, yes, of course. Will include that in a commit.

r0x0d commented:
@kokesak already added the ability to see what packages need to update in the log message. Now I'm taking a closer look at the repo thing.

r0x0d commented:
@kokesak I've fixed the scenario where it could have a duplicate repository defined in any repofiles under `/etc/yum.repos.d`. 

I think this is a corner case that should not happen very often, but nonetheless, I updated my parser for repoquery and now I'm throwing away that specific line!

r0x0d commented:
> @r0x0d Thanks for the fix. I did a little investigating on still failing OL8 issue. The error is this:
> 
> ```
>  out: [03/15/2022 18:15:13] TASK - [Prepare: Checking if the packages are up-to-date] *****************
> 18:15:13             out: Repository copr:copr.fedorainfracloud.org:group_oamg:convert2rhel is listed more than once in the configuration
> 18:15:13             out: Repository copr:copr.fedorainfracloud.org:group_oamg:convert2rhel is listed more than once in the configuration
> 18:15:14             out: Errors during downloading metadata for repository 'ol8_baseos_latest':
> 18:15:14             out:   - Curl error (6): Couldn't resolve host name for https://yum$ociregion.$ocidomain/repo/OracleLinux/OL8/baseos/latest/x86_64/repodata/repomd.xml [Could not resolve host: yum$ociregion.$ocidomain]
> 18:15:14             out: [03/15/2022 18:15:14] DEBUG - Traceback (most recent call last):
> 18:15:14             out:   File "/usr/lib/python3.6/site-packages/dnf/repo.py", line 574, in load
> 18:15:14             out:     ret = self._repo.load()
> 18:15:14             out:   File "/usr/lib64/python3.6/site-packages/libdnf/repo.py", line 397, in load
> 18:15:14             out:     return _repo.Repo_load(self)
> 18:15:14             out: libdnf._error.Error: Failed to download metadata for repo 'ol8_baseos_latest': Cannot download repomd.xml: Cannot download repodata/repomd.xml: All mirrors were tried
> 18:15:14             out: 
> 18:15:14             out: During handling of the above exception, another exception occurred:
> 18:15:14             out: 
> 18:15:14             out: Traceback (most recent call last):
> 18:15:14             out:   File "/usr/lib/python3.6/site-packages/convert2rhel/main.py", line 92, in main
> 18:15:14             out:     checks.perform_pre_checks()
> 18:15:14             out:   File "/usr/lib/python3.6/site-packages/convert2rhel/checks.py", line 60, in perform_pre_checks
> 18:15:14             out:     check_package_updates()
> 18:15:14             out:   File "/usr/lib/python3.6/site-packages/convert2rhel/checks.py", line 466, in check_package_updates
> 18:15:14             out:     packages_to_update = get_total_packages_to_update()
> 18:15:14             out:   File "/usr/lib/python3.6/site-packages/convert2rhel/pkghandler.py", line 986, in get_total_packages_to_update
> 18:15:14             out:     packages = _get_packages_to_update_dnf()
> 18:15:14             out:   File "/usr/lib/python3.6/site-packages/convert2rhel/pkghandler.py", line 1019, in _get_packages_to_update_dnf
> 18:15:14             out:     base.fill_sack()
> 18:15:14             out:   File "/usr/lib/python3.6/site-packages/dnf/base.py", line 399, in fill_sack
> 18:15:14             out:     self._add_repo_to_sack(r)
> 18:15:14             out:   File "/usr/lib/python3.6/site-packages/dnf/base.py", line 139, in _add_repo_to_sack
> 18:15:14             out:     repo.load()
> 18:15:14             out:   File "/usr/lib/python3.6/site-packages/dnf/repo.py", line 581, in load
> 18:15:14             out:     raise dnf.exceptions.RepoError(str(e))
> 18:15:14             out: dnf.exceptions.RepoError: Failed to download metadata for repo 'ol8_baseos_latest': Cannot download repomd.xml: Cannot download repodata/repomd.xml: All mirrors were tried
> 18:15:14             out: 
> ```
> 
> I think there is some problem when expending variables `$ociregion.$ocidomain`. I'm not sure if the same variables are being used on OL7 images. I will try to check it, but the conversions are ok on OL7.
> 
> I tried to look here https://docs.oracle.com/en-us/iaas/oracle-linux/getting-started/install-software-on-instance.htm The variable in /etc/yum/vars/ocidomain on the tested system conatins "oracle.com", the ociregion is empty. I wasn't able to find out more

@kokesak, I have made a fix for this issue! Thanks for reporting this out. I did see this yesterday when testing out but hadn't the proper time to investigate. 

Now it should work fine for OracleLinux 8.

r0x0d commented on convert2rhel/pkghandler.py:
```diff
@@ -933,3 +934,102 @@ def has_duplicate_repos_across_disablerepo_enablerepo_options():
+    base = Base()
+    # Fix the case when we are trying to query packages in Oracle Linux 8
+    # when the DNF API gets called, those variables are not populated by default.
+    if system_info.id == "oracle":
+        base.conf.substitutions["ocidomain"] = "oracle.com"
+        base.conf.substitutions["ociregion"] = ""
```
@SpyTec, this fix is for the issue that @kokesak raised (https://github.com/oamg/convert2rhel/pull/356#issuecomment-1069473404) 

Do you think it should be better for us to read the contents of /etc/yum/vars/{ocidomain,ociregion} instead of applying the defaults ourselves? I mean, the code could be running inside a machine that the user has changed the contents of the files, and then, our code would stop working. 

This only happens on OL8 as DNF does not expand those custom variables. (yum works fine, tho)

kokesak commented:
/packit test

kokesak commented on tests/integration/tier1/system-up-to-date/test_system_up_to_date.py:
```diff
@@ -0,0 +1,89 @@
+            env.str("RHSM_POOL"),
+        )
+    ) as c2r:
+        # TODO do we want to test the scenario where we have a package locked by this?
```
We could also run full conversion when the versionlock is applied. I would run together with some non updated package.

danmyway commented on tests/integration/tier1/set-locale/use_non_english_language.py:
```diff
@@ -15,3 +15,9 @@ def test_use_non_english_language(shell):
     assert shell("localectl list-locales | grep zh_CN.utf8").returncode == 0
     assert shell("localectl set-locale LANG=zh_CN.utf8").returncode == 0
+
+    # When running tests on TF on CentOS7 we are getting an error
+    # about setting incompatible LC_CTYPE C.UTF-8).
+    # However the C.UTF-8 was added on RHEL-8 like distros.
+    if "centos-7" in os:
+        assert shell("localectl set-locale LC_CTYPE=zh_CN.utf8").returncode == 0
```
```suggestion
    # Testing farm is returning an error on CentOS7 mentioning 
    # setting incompatible LC_CTYPE C.UTF-8.
    # However the C.UTF-8 was added on RHEL-8 like distros.
    if "centos-7" in os:
        assert os.system("localectl set-locale LC_CTYPE=zh_CN.utf8") == 0
```

danmyway commented on tests/integration/tier1/system-up-to-date/install_non_latest_kernel.py:
```diff
@@ -0,0 +1,24 @@
+    """
+    system_version = platform.platform()
+
+    # set deafault kernel
+    if "centos-7" in system_version:
+        assert shell("yum install kernel-3.10.0-1160.el7.x86_64 -y").returncode == 0
+        shell("grub2-set-default 'CentOS Linux (3.10.0-1160.el7.x86_64) 7 (Core)'")
+    elif "oracle-7" in system_version:
+        assert shell("yum install kernel-3.10.0-1160.el7.x86_64 -y").returncode == 0
+        shell("grub2-set-default 'Oracle Linux Server 7.9, with Linux 3.10.0-1160.el7.x86_64'")
+    elif "centos-8" in system_version:
+        assert shell("yum install kernel-4.18.0-240.22.1.el8_3 -y").returncode == 0
+        shell("grub2-set-default 'CentOS Linux (4.18.0-240.22.1.el8_3.x86_64) 8'")
+    elif "oracle-8" in system_version:
+        assert shell("yum install kernel-4.18.0-240.22.1.el8_3 -y").returncode == 0
+        shell("grub2-set-default 'Oracle Linux Server (4.18.0-240.22.1.el8_3.x86_64) 8.3'")
```
```suggestion
    # Set default kernel
    if "centos-7" in system_version:
        assert os.system("yum install kernel-3.10.0-1160.el7.x86_64 -y") == 0
        os.system("grub2-set-default 'CentOS Linux (3.10.0-1160.el7.x86_64) 7 (Core)'")
    elif "oracle-7" in system_version:
        assert os.system("yum install kernel-3.10.0-1160.el7.x86_64 -y") == 0
        os.system("grub2-set-default 'Oracle Linux Server 7.9, with Linux 3.10.0-1160.el7.x86_64'")
    elif "centos-8" in system_version:
        assert os.system("yum install kernel-4.18.0-240.22.1.el8_3 -y") == 0
        os.system("grub2-set-default 'CentOS Linux (4.18.0-240.22.1.el8_3.x86_64) 8'")
    elif "oracle-8" in system_version:
        assert os.system("yum install kernel-4.18.0-240.22.1.el8_3 -y") == 0
        os.system("grub2-set-default 'Oracle Linux Server (4.18.0-240.22.1.el8_3.x86_64) 8.3'")
```

danmyway commented on tests/integration/tier1/system-up-to-date/test_non_latest_kernel_inhibitor.py:
```diff
@@ -0,0 +1,47 @@
+    # We need to dynamically get the the name of the latest kernel
+    # that is in the repositories
+
+    # repoquery command requires 'yum-utils'
+    shell("yum install yum-utils -y")
```
```suggestion
    # Install 'yum-utils' required by the repoquery command
    os.system("yum install yum-utils -y")
```

danmyway commented on tests/integration/tier1/system-up-to-date/test_non_latest_kernel_inhibitor.py:
```diff
@@ -0,0 +1,47 @@
+    latest_kernel = shell(
+        "repoquery --quiet --qf '%{BUILDTIME}\t%{VERSION}-%{RELEASE}' kernel 2>/dev/null | tail -n 1 | awk '{printf $NF}'"
+    ).output
+    shell("echo {}".format(latest_kernel))
```
```suggestion
    os.system("echo {}".format(latest_kernel))
```

danmyway commented on tests/integration/tier1/system-up-to-date/test_non_latest_kernel_inhibitor.py:
```diff
@@ -0,0 +1,47 @@
+        "grubby --info ALL | grep \"title=.*{}\" | tr -d '\"' | sed 's/title=//'".format(latest_kernel)
+    ).output
+
+    # Set the latest kernel as one we want to reboot to
+    shell("grub2-set-default '{}'".format(full_name.strip()))
```
```suggestion
    # Set the latest kernel as the one we want to reboot to
    os.system("grub2-set-default '{}'".format(full_name.strip()))
```

danmyway commented on tests/integration/tier1/system-up-to-date/test_non_latest_kernel_inhibitor.py:
```diff
@@ -0,0 +1,47 @@
+import platform
```
```suggestion
import platform
import os
```

danmyway replied:
In case the `os.system` will be used instead of calling the `shell()` function

danmyway commented on tests/integration/tier1/system-up-to-date/test_system_up_to_date.py:
```diff
@@ -0,0 +1,103 @@
+    """
+    centos_8_pkg_url = "https://vault.centos.org/8.1.1911/BaseOS/x86_64/os/Packages/wpa_supplicant-2.7-1.el8.x86_64.rpm"
+
+    # Try to downgrade two packages
+    # On CentOS-8 we cannot do downgrade as the repos contains only the latest package version.
+    # To workaround this we need to install package from older repo.
+    if "centos-8" in system_version:
+        assert shell("yum install -y {}".format(centos_8_pkg_url)).returncode == 0
+    else:
+        assert shell("yum install openldap wpa_supplicant -y").returncode == 0
+        assert shell("yum downgrade openldap wpa_supplicant -y").returncode == 0
+
+    assert shell("yum install -y yum-plugin-versionlock").returncode == 0
+    assert shell("yum versionlock wpa_supplicant").returncode == 0
```
```suggestion
    # Try to downgrade two packages
    # On CentOS-8 we cannot do the downgrade as the repos contain only the latest package version.
    # We need to install package from older repository as a workaround.
    if "centos-8" in system_version:
        assert os.system("yum install -y {}".format(centos_8_pkg_url)) == 0
    else:
        assert os.system("yum install openldap wpa_supplicant -y") == 0
        assert os.system("yum downgrade openldap wpa_supplicant -y") == 0

    assert os.system("yum install -y yum-plugin-versionlock") == 0
    assert os.system("yum versionlock wpa_supplicant") == 0
```

danmyway commented on tests/integration/tier1/system-up-to-date/test_system_up_to_date.py:
```diff
@@ -0,0 +1,103 @@
+        c2r.sendline("n")
+
+    # Clean up
+    assert shell("yum versionlock clear").returncode == 0
```
```suggestion
    assert os.system("yum versionlock clear") == 0
```

danmyway commented on tests/integration/tier1/system-up-to-date/test_system_up_to_date.py:
```diff
@@ -0,0 +1,103 @@
+    kernel check inhibitor. One of the way to allow this is to not have
+    any kernel packages present in repos.
+    """
+    shell("mkdir /tmp/my_tmp; mv /etc/yum.repos.d/* /tmp/my_tmp")
```
```suggestion
    os.systeml("mkdir /tmp/my_tmp; mv /etc/yum.repos.d/* /tmp/my_tmp")
```

danmyway commented on tests/integration/tier1/system-up-to-date/test_system_up_to_date.py:
```diff
@@ -0,0 +1,103 @@
+def test_skip_kernel_check(shell, convert2rhel):
+    """
+    Test that it is possible to use env variable in some case to override
+    kernel check inhibitor. One of the way to allow this is to not have
```
```suggestion
    the kernel check inhibitor. One of the way to allow this is to not have
```

danmyway commented on tests/integration/tier1/system-up-to-date/test_system_up_to_date.py:
```diff
@@ -0,0 +1,103 @@
+    assert c2r.exitstatus != 0
+
+    # Clean up
+    shell("mv /tmp/my_tmp/* /etc/yum.repos.d/")
```
```suggestion
    os.system("mv /tmp/my_tmp/* /etc/yum.repos.d/")
```

danmyway reviewed (changes requested):
Mainly minor typo and stylistic corrections.
I would try to use `os.system()` instead of calling `shell()` function, whenever the output is not needed.

kokesak commented on plans/tier1.fmf:
```diff
@@ -204,3 +204,26 @@ discover+:
+/system_up_to_date:
+  discover+:
+    test: checks-after-conversion
+  prepare+:
+  - name: preapre non latest kernel
+    how: shell
+    script: pytest -svv tests/integration/tier1/system-up-to-date/install_non_latest_kernel.py
+  - name: reboot machine
+    how: ansible
+    playbook: tests/ansible_collections/roles/reboot/main.yml
+  - name: test inhibitor on non latest kernels
+    how: shell
+    script: pytest -svv tests/integration/tier1/system-up-to-date/test_non_latest_kernel_inhibitor.py
+  - name: reboot machine
+    how: ansible
+    playbook: tests/ansible_collections/roles/reboot/main.yml
+  - name: test conversion non updated package
+    how: shell
+    script: pytest -svv tests/integration/tier1/system-up-to-date/test_system_up_to_date.py
+  - name: reboot after conversion
+    how: ansible
+    playbook: tests/ansible_collections/roles/reboot/main.yml
```
Sadly no. We need to reboot the machine between some steps and the reboot may be tricky to implement correctly in the test itself.

danmyway replied:
Oh, alright then!

kokesak commented on tests/integration/tier0/backup-release/test_backup_release.py:
```diff
@@ -24,6 +24,9 @@ def test_backup_os_release_no_envar(shell, convert2rhel):
     # Move all repos to other location so it is not used
```
Thanks for catching the typoos in comments!

abadger commented on tests/integration/tier0/backup-release/test_backup_release.py:
```diff
@@ -24,6 +24,9 @@ def test_backup_os_release_no_envar(shell, convert2rhel):
     # Move all repos to other location so it is not used
     assert shell("mkdir /tmp/s_backup && mv /etc/yum.repos.d/* /tmp/s_backup/").returncode == 0
```
kokesak, danmyway, and I talked about this today and we decided it was okay to stick with shell.  subprocess is the encouraged way to call out to other programs ( First paragraph of https://docs.python.org/3/library/subprocess.html ).  The other functions in the python stdlib are considered legacy ways of calling out.

The one thing about the `shell()` function to be aware of is that (as the name implies) it sends the command to the shell.  This means that shell metacharacters are processed by the shell (`*`, `;`, `&&`, etc).  This can be a source of security bugs whenever user input is used but I don't believe that's a problem for our integration tests in general.

kokesak replied:
@abadger thanks, we will stick with the `shell()` function.

danmyway reviewed (approved):
LGTM! Thank you!

SpyTec commented on convert2rhel/pkghandler.py:
```diff
@@ -968,3 +969,109 @@ def get_pkg_names_from_rpm_paths(rpm_paths):
+    # Same thing for CentOS, but this time, the URL for the vault has changed
+    # and instead of just ussing $releasever (8.5, 8.4...), we need either
+    # using the correct $releasever (8.5.2111) or setting up the $contentdir
+    if system_info.id == "centos":
```
```suggestion
    elif system_info.id == "centos":
```

SpyTec reviewed (commented):
@r0x0d I skimmed and saw two comments about some pieces of code not being tested, which has to do with branches. The `sys.path.insert(0, "/usr/share/yum-cli")` was never asserted to be called, and `base.conf.substitutions` was never asserted either

r0x0d commented:
Hi, @abadger and @bocekm could one of you do a final review & approval in this PR? Thanks!

bocekm commented on convert2rhel/pkghandler.py:
```diff
@@ -968,3 +969,109 @@ def get_pkg_names_from_rpm_paths(rpm_paths):
+    # Fix the case when we are trying to query packages in Oracle Linux 8
+    # when the DNF API gets called, those variables are not populated by default.
+    if system_info.id == "oracle":
+        base.conf.substitutions["ocidomain"] = "oracle.com"
```
Instead of hardcoding the value, it'd better to load the variables from the `/etc/dnf/vars/`. Oracle introduced this $ocidomain variable in
https://yum.oracle.com/repo/OracleLinux/OL8/baseos/latest/x86_64/getPackage/[oraclelinux-release-el8-1.0-16.0.1.el8.x86_64.rpm](https://yum.oracle.com/repo/OracleLinux/OL8/baseos/latest/x86_64/getPackage/oraclelinux-release-el8-1.0-16.0.1.el8.x86_64.rpm). 

You can change it in a subsequent PR.

Related comment: https://github.com/oamg/convert2rhel/pull/356#issuecomment-1069381290

r0x0d replied:
Bugzilla from someone that had "the same problem": https://bugzilla.redhat.com/show_bug.cgi?id=1920735

r0x0d replied:
Fixed in https://github.com/oamg/convert2rhel/pull/458/files#r864814726
